src/script/sign.cpp

```diff
@@ -302,6 +381,218 @@ struct Stacks
+        {
+            CScript witnessscript;
+            witnessscript << OP_DUP << OP_HASH160 << ToByteVector(script_ret[0]) << OP_EQUALVERIFY << OP_CHECKSIG;
+            txnouttype subType;
+            solved = solved && SignSigsOnly(creator, witnessscript, sig_ret, subType, SIGVERSION_WITNESS_V0, key_ret, script_ret, psbt.redeem_scripts, psbt.witness_scripts);
+        }
+        else if (solved && whichType == TX_WITNESS_V0_SCRIPTHASH)
+        {
+            CScript witnessscript(script_ret[0].begin(), script_ret[0].end());
+            txnouttype subType;
+            solved = solved && SignSigsOnly(creator, witnessscript, sig_ret, subType, SIGVERSION_WITNESS_V0, key_ret, script_ret, psbt.redeem_scripts, psbt.witness_scripts) && subType != TX_SCRIPTHASH && subType != TX_WITNESS_V0_SCRIPTHASH && subType != TX_WITNESS_V0_KEYHASH;
+        }
+
+        // Add to partial sigs
+        if (solved) {
+            for (unsigned int j = 0; j < key_ret.size(); ++j) {
+                psbt.inputs[i].partial_sigs.emplace(key_ret[j], sig_ret[j]);
+            }
+        }
+    }
+
+    return solved;
+}
+
+// Finalizes the inputs that can be finalized
+// Returns true for final tx, false for non final
+bool FinalizePartialTransaction(PartiallySignedTransaction& psbt)
+{
+    CMutableTransaction mtx = psbt.tx;
+    bool complete = true;
+    const CTransaction const_tx(mtx);
+    for (unsigned int i = 0; i < mtx.vin.size(); ++i) {
+        CTxIn& txin = mtx.vin[i];
+        PartiallySignedInput psbt_in = psbt.inputs[i];
+
+        // Find the non witness utxo first
+        CTxOut utxo;
+        if (psbt_in.non_witness_utxo) {
+            utxo = psbt_in.non_witness_utxo->vout[txin.prevout.n];
+        }
+        // Now find the witness utxo if the non witness doesn't exist
+        else if (!psbt_in.witness_utxo.IsNull()) {
+            utxo = psbt_in.witness_utxo;
+        }
+        // If there is no nonwitness or witness utxo, then this input is fully signed and we are done here
+        else {
+            continue;
+        }
+
+        // Combine partial sigs and create full scriptsig
+        std::vector<valtype> vSolutions;
+        CScript spk = utxo.scriptPubKey;
+        bool loop = true;
+        bool P2SH = false;
+        bool witness = false;
+        bool WSH = false;
+        CScript redeemscript;
+        CScript witnessscript;
+        SignatureData sigdata;
+        std::vector<valtype> script_ret; // Only for the signer to put redemscripts to be used later.
+        txnouttype whichType;
+        while (loop) {
+            loop = false;
+            uint160 h160;
+            CScript script_ret2;
+            CKeyID keyID;
+            bool found_pk = false;
+            if (Solver(spk, whichType, vSolutions)) {
+                switch (whichType)
+                {
+                case TX_PUBKEY:
+                    script_ret.push_back(psbt.inputs[i].partial_sigs.find(CPubKey(vSolutions[0]))->second);
```

Comment:
this may segfault when there is no corresponding `partial_sig`, I believe this is the cause of your test failure.

Reply:
Fixed